controller/konnect/reconciler_kongplugin.go

```diff
@@ -88,130 +103,124 @@ func (r *KongPluginReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	// the same logic for KongRoute, KongConsumer, and KongConsumerGroup as well.
 	// https://github.com/Kong/gateway-operator/issues/583
 
-	// Group the PluginBindings by reference type and name.
-	bindingMapping := mapKongPluginBindingsByTargetTypeAndRef(kongPluginBindingList.Items)
-
-	// Get all the KongServices referenced by the KongPluginBindings
-	// TODO(mlavacca): use indexers instead of listing all KongServices
-	// https://github.com/Kong/gateway-operator/issues/596
 	kongServiceList := configurationv1alpha1.KongServiceList{}
-	err = r.client.List(
-		ctx,
-		&kongServiceList,
-		client.InNamespace(kongPlugin.Namespace),
+	err = clientWithNamespace.List(ctx, &kongServiceList,
+		client.MatchingFields{
+			IndexFieldKongServiceOnReferencedPluginNames: kongPlugin.Namespace + "/" + kongPlugin.Name,
+		},
 	)
 	if err != nil {
 		return ctrl.Result{}, err
 	}
 
-	pluginBindingsToDelete := []configurationv1alpha1.KongPluginBinding{}
-	// pluginReferenceFound represents whether the plugin is referenced by any KongService
+	kongRouteList := configurationv1alpha1.KongRouteList{}
+	err = clientWithNamespace.List(ctx, &kongRouteList,
+		client.MatchingFields{
+			IndexFieldKongRouteOnReferencedPluginNames: kongPlugin.Namespace + "/" + kongPlugin.Name,
+		},
+	)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	foreignRelations := ForeignRelations{
+		Service: lo.Filter(kongServiceList.Items,
+			func(s configurationv1alpha1.KongService, _ int) bool { return s.DeletionTimestamp.IsZero() },
+		),
+		Route: lo.Filter(kongRouteList.Items,
+			func(s configurationv1alpha1.KongRoute, _ int) bool { return s.DeletionTimestamp.IsZero() },
+		),
+		// TODO consumers and consumer groups
+	}
+	grouped, err := foreignRelations.GroupByControlPlane(ctx, r.client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// pluginReferenceFound represents whether the plugin is referenced by any resource.
 	var pluginReferenceFound bool
```

Comment:
Please move this declaration just before the `for cpNN, relations := range groupedCombinations` loop (to keep it as close to its usage as possible).

Reply:
fc141e0f75fb49b8623b1c04d37eaa81587ee419